utils/fasm/src/main.cpp

```diff
@@ -0,0 +1,144 @@
+/**
+VPR is a CAD tool used to conduct FPGA architecture exploration.  It takes, as input, a technology-mapped netlist and a description of the FPGA architecture being investigated.
+ VPR then generates a packed, placed, and routed FPGA (in .net, .place, and .route files respectively) that implements the input netlist.
+
+ This file is where VPR starts execution.
+
+ Key files in VPR:
+ 1.  libarchfpga/physical_types.h - Data structures that define the properties of the FPGA architecture
+ 2.  vpr_types.h - Very major file that defines the core data structures used in VPR.  This includes detailed architecture information, user netlist data structures, and data structures that describe the mapping between those two.
+ 3.  globals.h - Defines the global variables used by VPR.
+ */
+
+#include <cstdio>
+#include <cstring>
+#include <ctime>
+#include <fstream>
+using namespace std;
+
+#include "vtr_error.h"
+#include "vtr_memory.h"
+#include "vtr_log.h"
+
+#include "tatum/error.hpp"
+
+#include "vpr_error.h"
+#include "vpr_api.h"
+#include "vpr_signal_handler.h"
+
+#include "globals.h"
+
+#include "net_delay.h"
+#include "RoutingDelayCalculator.h"
+
+#include "fasm.h"
+
+/*
+ * Exit codes to signal success/failure to scripts
+ * calling vpr
+ */
+constexpr int SUCCESS_EXIT_CODE = 0; //Everything OK
+constexpr int ERROR_EXIT_CODE = 1; //Something went wrong internally
+constexpr int UNIMPLEMENTABLE_EXIT_CODE = 2; //Could not implement (e.g. unroutable)
+constexpr int INTERRUPTED_EXIT_CODE = 3; //VPR was interrupted by the user (e.g. SIGINT/ctr-C)
+
+static bool write_fasm() {
+  auto& atom_ctx = g_vpr_ctx.atom();
+
+  std::string fasm_filename = atom_ctx.nlist.netlist_name() + ".fasm";
+  vtr::printf("Writing Implementation FASM: %s\n", fasm_filename.c_str());
+  std::ofstream fasm_os(fasm_filename);
+  FasmWriterVisitor visitor2(fasm_os);
+  NetlistWalker nl_walker2(visitor2);
+  nl_walker2.walk();
+
+  return true;
+}
+
+/**
+ * VPR program
+ * Generate FPGA architecture given architecture description
+ * Pack, place, and route circuit into FPGA architecture
+ * Electrical timing analysis on results
+ *
+ * Overall steps
+ * 1.  Initialization
+ * 2.  Pack
+ * 3.  Place-and-route and timing analysis
+ * 4.  Clean up
+ */
+int main(int argc, const char **argv) {
+    t_options Options = t_options();
+    t_arch Arch = t_arch();
+    t_vpr_setup vpr_setup = t_vpr_setup();
+    clock_t entire_flow_begin, entire_flow_end;
+
+    entire_flow_begin = clock();
+
+    try {
+        vpr_install_signal_handler();
+
+        /* Read options, architecture, and circuit netlist */
+        vpr_init(argc, argv, &Options, &vpr_setup, &Arch);
+
+        vpr_setup.PackerOpts.doPacking    = STAGE_LOAD;
+        vpr_setup.PlacerOpts.doPlacement  = STAGE_LOAD;
+        vpr_setup.RouterOpts.doRouting    = STAGE_LOAD;
+        vpr_setup.AnalysisOpts.doAnalysis = STAGE_SKIP;
```

Comment:
Is there a `STAGE_LOAD` for analysis at all?

Reply:
We don't need it for FASM output?